contracts/src/libraries/upgradeable.cairo

```diff
@@ -11,8 +11,15 @@ mod Upgradeable {
     #[event]
     fn Upgraded(implementation: ClassHash) {}
 
-    fn upgrade(impl_hash: ClassHash) {
+    fn upgrade_only_owner(impl_hash: ClassHash) {
         Ownable::assert_only_owner();
+        upgrade(impl_hash);
+    }
+
+    // this method assumes replace_class_syscall has a very low possibility of being deprecated
+    // but if it does, we will either have upgraded the contract to be non-upgradeable by then
+    // because the starknet ecosystem has stabilized or we will be able to upgrade the contract to the proxy pattern
+    fn upgrade(impl_hash: ClassHash) {
```

Comment:
This seems dangerous as the caller might forget to add access controls or expect they are already in place.

Can we have just one `upgrade` entry point which internally always checks access by using an abstract fn like `assert_only_upgrade_admin()`, which then downstream contracts need to implement?

Ownable contracts can implement using `Ownable::assert_only_owner();` while others have their own preferred mechanism. 

The point is this will be enforced and explicit.

Reply:
Ok, sounds good. I couldn't find a better alternative so looks like we'll leave it as is. OZ Cairo lib implementations also use a similar pattern.